CMakeLists.txt

```diff
@@ -1818,6 +1818,74 @@ add_library_variants_for_cpu(
     RAM_SIZE 0x1000000
     STACK_SIZE 4K
 )
+# FIXME: qemu currently has no support for PACBTI-M, so the branch protection
+# variants below can't be fully tested in runtime. Since PACBTI-M instructions
+# are NOP-compatible we can use the cortext-m55 CPU for now, but these should be
+# updated to use a PACBTI-M enabled CPU once this is available.
+add_library_variants_for_cpu(
+    armv8.1m.main
+    SUFFIX soft_nofp_nomve_pacret_bti
+    COMPILE_FLAGS "-mfloat-abi=soft -march=armv8.1m.main+nomve -mfpu=none -mbranch-protection=standard"
+    MULTILIB_FLAGS "--target=thumbv8.1m.main-unknown-none-eabi -mfpu=none -mbranch-protection=standard"
+    PICOLIBC_BUILD_TYPE "release"
+    QEMU_MACHINE "mps3-an547"
+    QEMU_CPU "cortex-m55"
+    BOOT_FLASH_ADDRESS 0x00000000
+    BOOT_FLASH_SIZE 512K
+    FLASH_ADDRESS 0x60000000
+    FLASH_SIZE 0x1000000
+    RAM_ADDRESS 0x61000000
+    RAM_SIZE 0x1000000
+    STACK_SIZE 4K
+)
+add_library_variants_for_cpu(
+    armv8.1m.main
+    SUFFIX hard_fp_nomve_pacret_bti
+    COMPILE_FLAGS "-mfloat-abi=hard -march=armv8.1m.main+nomve -mfpu=fp-armv8-fullfp16-sp-d16 -mbranch-protection=standard"
+    MULTILIB_FLAGS "--target=thumbv8.1m.main-unknown-none-eabihf -march=thumbv8.1m.main+fp16 -mfpu=fp-armv8-fullfp16-sp-d16 -mbranch-protection=standard"
```

Comment:
Why is the `+fp16` on the `-march` option needed here, but not in `COMPILE_FLAGS`?

Reply:
In `COMPILE_FLAGS` the `+fp16` extension is enabled by the choice of FPU.
The difference in the list of flags is due to the normalisation applied to the command line options for the library selection:
```
$ ../build/llvm/bin/clang --target=arm-none-eabi -mfloat-abi=hard -march=armv8.1m.main+nomve -mfpu=fp-armv8-fullfp16-sp-d16 -mbranch-protection=standard test.c -print-multi-flags-experimental
--target=thumbv7em-unknown-none-eabihf
--target=thumbv7m-unknown-none-eabihf
--target=thumbv8.1m.main-unknown-none-eabihf
--target=thumbv8m.main-unknown-none-eabihf
-fexceptions
-frtti
-march=thumbv8.1m.main+fp16
-march=thumbv8.1m.main+fp16+nosha2+noaes+nomve+nomve.fp
-mbranch-protection=pac-ret+bti
-mbranch-protection=standard
-mfloat-abi=hard
-mfpu=fp-armv8-fullfp16-sp-d16
-mfpu=fpv4-sp-d16
-mfpu=fpv5-sp-d16
-mfpu=vfpv3xd
-mfpu=vfpv3xd-fp16
```